From 9e024bdb939a98c437ce9f5875005ca407e81823 Mon Sep 17 00:00:00 2001
From: John Zhuge <jzhuge@cloudera.com>
Date: Fri, 26 May 2017 10:13:45 -0700
Subject: [PATCH 2467/2748] HDFS-11709. StandbyCheckpointer should handle an
 non-existing legacyOivImageDir gracefully.
 Contributed by Erik Krogen.

(cherry picked from commit d8a33098309f17dfb0e3a000934f68394de44bf7)
(cherry picked from commit fc46f398d9192e83c7197115589eca7805c49943)
(cherry picked from commit 70f18911c71c61f9b3918e195bfb9a99d22d230c)
(cherry picked from commit bec23593b55e6b35366a1d804d0fb335d954ad6a)

Change-Id: Icd6cbaeb0951f0b4c3aa9e4cf1ab335daf480c9e
(cherry picked from commit 2086fd17cbe016641edeb497270dd52c2b4727d3)
---
 .../server/namenode/ha/StandbyCheckpointer.java    |    7 ++++++-
 1 file changed, 6 insertions(+), 1 deletion(-)

diff --git a/hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/server/namenode/ha/StandbyCheckpointer.java b/hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/server/namenode/ha/StandbyCheckpointer.java
index 31fa5a2..5c09043 100644
--- a/hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/server/namenode/ha/StandbyCheckpointer.java
+++ b/hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/server/namenode/ha/StandbyCheckpointer.java
@@ -190,7 +190,12 @@ private void doCheckpoint() throws InterruptedException, IOException {
       // Save the legacy OIV image, if the output dir is defined.
       String outputDir = checkpointConf.getLegacyOivImageDir();
       if (outputDir != null && !outputDir.isEmpty()) {
-        img.saveLegacyOIVImage(namesystem, outputDir, canceler);
+        try {
+          img.saveLegacyOIVImage(namesystem, outputDir, canceler);
+        } catch (IOException ioe) {
+          LOG.error("Exception encountered while saving legacy OIV image; "
+                  + "continuing with other checkpointing steps", ioe);
+        }
       }
     } finally {
       namesystem.cpUnlock();
-- 
1.7.9.5

